### PR TITLE
Using s390x vsis for Peer-pod demo

### PR DESCRIPTION
**Pull request by liudalibj**
- update cloud-api-adaptor to support IBM test cloud env
- update ibmcloud/terraform/cluster/ansible/playbook.yml to support build pod vm s390x image on work node
- add document about `using s390x vsi for Peer-pod demo`
Signed-off by: Liu Da Li <liudali@cn.ibm.com>

### CODE REVIEW

File: ibmcloud/terraform/cluster/ansible/playbook.yml [hunk 1]
```diff
@@ -308,7 +311,7 @@
       shell: |
         set -o errexit -o pipefail
         cd /root
-        git clone -b peerpod-2022.04.04 https://github.com/yoheiueda/cloud-api-adaptor.git
+        git clone -b zvsi https://github.com/liudalibj/cloud-api-adaptor.git
```
Comment (stevenhorsman) [inline ibmcloud/terraform/cluster/ansible/playbook.yml:314]: @liudalibj, @yoheiueda - Hi Da Li & Yohei, I wonder what the plan is once this PR merges.  Would we consider creating a new tag and then update this playbook to pull that version, or we could keep Da Li's branch until we are ready to contribute the whole `code-import` branch in Yohei's fork to `staging` (which might be around the end of the week)?

Comment (liudalibj) [reply ibmcloud/terraform/cluster/ansible/playbook.yml:314]: I would like create a new tag by @yoheiueda after this PR be merged to `code-import` and those url be changed back to yoheiueda repo.

File: ibmcloud/README-ZVSI.md [hunk 2]
```diff
@@ -0,0 +1,350 @@
+# Setup procedure for IBM TEST Cloud (Staging environment)
```
Comment (stevenhorsman) [inline ibmcloud/README-ZVSI.md]: Given that the IBM Test Cloud information isn't widely relevant to open source and zVSI are creating on production I wonder if it is better to instead contribute some of the general fixes and improvements into README.md and add a little section on the differences e.g. in the Create a Kubernetes cluster section, after the instructions to add variables to the `terraform.tfvars` files you could add something like:
> **Note:** in order to create the cluster based on a different type of VSI image you can overwrite more parameters here e.g. to create an `s390x` based cluster add
>  ```
>  instance_profile_name = "bz2-2x8"
>  image_name = "ibm-ubuntu-18-04-1-minimal-s390x-3"
>  ```
That would stop us having to maintain two file with very similar content and still provide some zVSI instructions that should work?

Comment (liudalibj) [reply ibmcloud/README-ZVSI.md]: Agreed.
Merged some zVSI instructions to README.md so that we only need maintain one document

Comment (stevenhorsman) [reply ibmcloud/README-ZVSI.md]: Great. Thanks!

File: ibmcloud/README.md [hunk 3]
```diff
@@ -112,57 +158,79 @@ You need to build a pod VM image for peer pod VMs. A pod VM image contains the f
...
-The following environment variables are necessary to be set before executing the image upload script. You can change IBMCLOUD_COS_REGION if you prefer another region. In this case, you also want to change IBMCLOUD_COS_SERVICE_ENDPOINT to one of endpoints listed at [https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-endpoints](https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-endpoints).
+
+The following environment variables are necessary to be set before executing the image upload script. You can change `IBMCLOUD_COS_REGION` if you prefer another region. In this case, you also want to change `IBMCLOUD_COS_SERVICE_ENDPOINT` to one of endpoints listed at [https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-endpoints](https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-endpoints).
 
 ```
 export IBMCLOUD_API_KEY=<your API key>
 export IBMCLOUD_COS_SERVICE_INSTANCE=<COS service instance name>
 export IBMCLOUD_COS_BUCKET=<COS bucket name>
+export IBMCLOUD_API_ENDPOINT=https://cloud.ibm.com
+export IBMCLOUD_VPC_REGION=us-south
```
Comment (stevenhorsman) [inline ibmcloud/README.md]: I just wanted to check if `IBMCLOUD_VPC_REGION` should be `jp-tok` in this documentation to match the rest of the defaults?

File: ibmcloud/README.md [hunk 4]
```diff
@@ -19,12 +19,37 @@ The setup procedure includes the following sub tasks.
 To automate preparation of VPC and VSIs, you need to install terraform and ansible on your client machine. Please follow the the official installation guides.
 
 * [Install Terraform](https://learn.hashicorp.com/tutorials/terraform/install-cli)
+
+
+Simply:
```
Comment (mattarnoatibm) [inline ibmcloud/README.md]: Do we want to mark these instructions as instructions that will only work on (I assume) Ubuntu, or a Linux distro that uses the apt package manager.

I assume the instructions in this README can be ran on any client machine; Windows, Mac OS, any Linux distro, which is why the instructions for installing Terraform and Ansible point at the official guides from hashicorp and the Ansible project

**Review by mattarnoatibm — COMMENTED** [review]
I left a comment about the Terraform and Ansible installation instructions.

File: ibmcloud/README.md (hunk 3 above)
Comment (liudalibj) [reply ibmcloud/README.md]: Thanks, a typo, fixed.

File: ibmcloud/README.md (hunk 4 above)
Comment (liudalibj) [reply ibmcloud/README.md]: Thanks, replace `Simply` to `If you are using Ubuntu linux, you can run follow commands simply`

**Comment by stevenhorsman** [comment]
As a data point, rather than suggesting the PR is wrong I've hit a problem with the demo. When the peer pod VSI is created it fails to start due to capacity problems:
image
I'm not sure if this is just zVSI's being flakey or not, but as I created the cluster VSIs yesterday and the peer pod one today and the instructions have changed since then I think I'll destroy what I have and start clean and see if I get a different result.

**Comment by liudalibj** [comment]
> As a data point, rather than suggesting the PR is wrong I've hit a problem with the demo. When the peer pod VSI is created it fails to start due to capacity problems: image I'm not sure if this is just zVSI's being flakey or not, but as I created the cluster VSIs yesterday and the peer pod one today and the instructions have changed since then I think I'll destroy what I have and start clean and see if I get a different result.

Hello @stevenhorsman you don't need destroy your cluster, you can try this option:
Stop the running cloud-api-adaptor, start it again with another `vpc_zone` maybe `jp-tok-1` or `jp-tok-3` and then try create nginx demo again.

**Comment by stevenhorsman** [comment]
> Hello @stevenhorsman you don't need destroy your cluster, you can try this option: Stop the running cloud-api-adaptor, start it again with another `vpc_zone` maybe `jp-tok-1` or `jp-tok-3` and then try create nginx demo again.

Thanks, I'll give that a go. Is there an issue with the peer pod VSI running on a different zone to the cluster?

**Comment by liudalibj** [comment]
> Thanks, I'll give that a go. Is there an issue with the peer pod VSI running on a different zone to the cluster?

It should be no issues, peer pod VSI is controlled by `cloud-api-adaptor` on worker node,  not related to the cluster itself.

**Comment by stevenhorsman** [comment]
> It should be no issues, peer pod VSI is controlled by `cloud-api-adaptor` on worker node, not related to the cluster itself.

Hmm, I've done something else wrong then as when I start the cloud-api-adaptor I get the following error:
```
"Result": {
        "errors": [
            {
                "code": "bad_field",
                "message": "the network interface's subnet is located in a different zone from the vsi",
                "target": {
                    "name": "primary_network_interface.subnet",
                    "type": "field"
                }
            }
        ],
        "trace": "96db56d2-42cc-4478-a351-c7c86907b47c"
    },
```
My subnet and cp and worker VSIs are running on Tokyo 3 and I'm trying to bring up the run the cloud-api-adaptor with `vpc_zone=jp-tok-2`

**Comment by liudalibj** [comment]
Sorry missed one step, please create a new subnet on the target zone by hand and then use it as `primary-subnet-id` when start `cloud-api-adaptor` and then try again.

**Comment by stevenhorsman** [comment]
Now I've switch zones I'm getting a different issue with storage when trying to start the peer pod zVSI:
image
I'll try and go back through the docs and see if I can spot something I've missed.

File: ibmcloud/README.md [hunk 5]
```diff
@@ -219,7 +287,7 @@ vpc_region=jp-tok
 vpc_zone=jp-tok-2
 instance_profile=bx2-2x8
```
Comment (stevenhorsman) [inline ibmcloud/README.md:288]: I wonder if it is worth having a s390x note here? Something like:
> **Note:** Modify `instance_profile` to change the type of VSI provisioned e.g. to create 2 vCPU, 8GB RAM balanced `s390x` VSIs for the peer-pod use `instance_profile=bz2-2x8`

**Comment by stevenhorsman** [comment]
> Now I've switched zones I'm getting a different issue with storage when trying to start the peer pod zVSI: <img 

I had a couple of hours of meetings and noticed when I came back that the VSI had started (after I'd already killed the cloud-api-adaptor process), so I re-launched the cloud-api-adaptor and created the pod and it all worked this time:
```
# kubectl get pods
NAME    READY   STATUS    RESTARTS   AGE
nginx   1/1     Running   0          66s
```
```
# ibmcloud is instances
Listing instances in all resource groups and region jp-tok under account Steven Horsman's Account as user steven@uk.ibm.com...
ID                                          Name                                       Status     Reserved IP    Floating IP       Profile   Image                                VPC       Zone       Resource group
02f7_6cca4369-3047-452c-a2ce-0088e7bfd0b1   peer-pod-z-worker-default-nginx-f605722c   starting   10.244.65.5    -                 bz2-2x8   podvm-32a2850-dirty-s390x            tok-vpc   jp-tok-2   default
02g7_5bcbd6e7-6862-416d-8af8-85e7fc7b1789   peer-pod-z-cp                              running    10.244.128.4   165.192.136.236   bz2-2x8   ibm-ubuntu-18-04-1-minimal-s390x-3   tok-vpc   jp-tok-3   default
02g7_21fdf0a4-ccca-4367-b942-04a7cf16c0f9   peer-pod-z-worker                          running    10.244.128.5   165.192.130.197   bz2-2x8   ibm-ubuntu-18-04-1-minimal-s390x-3   tok-vpc   jp-tok-3   default
```
```
# kubectl exec nginx -- uname -a
Linux nginx 5.4.0-109-generic #123-Ubuntu SMP Fri Apr 8 11:56:05 UTC 2022 s390x GNU/Linux
```

I'm not entirely sure what caused the storage issues, but they seem to have gone now, so probably platform related?

**Comment by liudalibj** [comment]
> > Now I've switched zones I'm getting a different issue with storage when trying to start the peer pod zVSI:  
> I had a couple of hours of meetings and noticed when I came back that the VSI had started (after I'd already killed the cloud-api-adaptor process), so I re-launched the cloud-api-adaptor and created the pod and it all worked this time:
> 
> ```
> # kubectl get pods
> NAME    READY   STATUS    RESTARTS   AGE
> nginx   1/1     Running   0          66s
> ```
> 
> ```
> # ibmcloud is instances
> Listing instances in all resource groups and region jp-tok under account Steven Horsman's Account as user steven@uk.ibm.com...
> ID                                          Name                                       Status     Reserved IP    Floating IP       Profile   Image                                VPC       Zone       Resource group
> 02f7_6cca4369-3047-452c-a2ce-0088e7bfd0b1   peer-pod-z-worker-default-nginx-f605722c   starting   10.244.65.5    -                 bz2-2x8   podvm-32a2850-dirty-s390x            tok-vpc   jp-tok-2   default
> 02g7_5bcbd6e7-6862-416d-8af8-85e7fc7b1789   peer-pod-z-cp                              running    10.244.128.4   165.192.136.236   bz2-2x8   ibm-ubuntu-18-04-1-minimal-s390x-3   tok-vpc   jp-tok-3   default
> 02g7_21fdf0a4-ccca-4367-b942-04a7cf16c0f9   peer-pod-z-worker                          running    10.244.128.5   165.192.130.197   bz2-2x8   ibm-ubuntu-18-04-1-minimal-s390x-3   tok-vpc   jp-tok-3   default
> ```
> 
> ```
> # kubectl exec nginx -- uname -a
> Linux nginx 5.4.0-109-generic #123-Ubuntu SMP Fri Apr 8 11:56:05 UTC 2022 s390x GNU/Linux
> ```
> 
> I'm not entirely sure what caused the storage issues, but they seem to have gone now, so probably platform related?

Agreed that there are some platform related issue for zVSIs mainly are resource related, we need retry on different zone sometimes.

**Comment by liudalibj** [comment]
close this PR, move it to https://github.com/confidential-containers/cloud-api-adaptor/pull/9

**Pull request closed without merging**